File: environment/src/main/java/jetbrains/exodus/gc/UtilizationProfile.java
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.gc;

import jetbrains.exodus.bindings.LongBinding;
import jetbrains.exodus.core.dataStructures.Pair;
import jetbrains.exodus.core.dataStructures.hash.LongHashMap;
import jetbrains.exodus.core.execution.Job;
import jetbrains.exodus.env.*;
import jetbrains.exodus.log.*;
import jetbrains.exodus.tree.LongIterator;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.lang.ref.WeakReference;
import java.util.*;

public final class UtilizationProfile {

    @NotNull
    private final EnvironmentImpl env;
    @NotNull
    private final GarbageCollector gc;
    @NotNull
    private final Log log;
    private final long fileSize; // in bytes
    @NotNull
    private final LongHashMap<MutableLong> filesUtilization; // file address -> number of free bytes
    private long totalBytes;
    private long totalFreeBytes;
    private volatile boolean isDirty;

    UtilizationProfile(@NotNull final EnvironmentImpl env, @NotNull final GarbageCollector gc) {
        this.env = env;
        this.gc = gc;
        log = env.getLog();
        fileSize = log.getFileSize() * 1024L;
        filesUtilization = new LongHashMap<>();
        log.addNewFileListener(new NewFileListener() {
            @Override
            @SuppressWarnings({"ConstantConditions"})
            public void fileCreated(long fileAddress) {
                synchronized (filesUtilization) {
                    filesUtilization.put(fileAddress, new MutableLong(0L));
                }
                estimateTotalBytes();
            }
        });
    }

    void clear() {
        synchronized (filesUtilization) {
            filesUtilization.clear();
        }
        estimateTotalBytes();
    }

    /**
     * Loads utilization profile.
     */
    public void load() {
        final EnvironmentConfig ec = env.getEnvironmentConfig();
        if (ec.getGcUtilizationFromScratch()) {
            computeUtilizationFromScratch();
        } else {
            final String storedUtilization = ec.getGcUtilizationFromFile();
            if (!storedUtilization.isEmpty()) {
                loadUtilizationFromFile(storedUtilization);
            } else {
                env.executeInReadonlyTransaction(new TransactionalExecutable() {
                    @Override
                    public void execute(@NotNull final Transaction txn) {
                        if (!env.storeExists(GarbageCollector.UTILIZATION_PROFILE_STORE_NAME, txn)) {
                            if (env.getAllStoreCount() == 0 && log.getNumberOfFiles() <= 1) {
                                clearUtilization();
                            } else {
                                computeUtilizationFromScratch();
                            }
                        } else {
                            final LongHashMap<MutableLong> filesUtilization = new LongHashMap<>();
                            final StoreImpl store = env.openStore(GarbageCollector.UTILIZATION_PROFILE_STORE_NAME, StoreConfig.WITHOUT_DUPLICATES, txn);
                            try (Cursor cursor = store.openCursor(txn)) {
                                while (cursor.getNext()) {
                                    final long fileAddress = LongBinding.compressedEntryToLong(cursor.getKey());
                                    final long freeBytes = CompressedUnsignedLongByteIterable.getLong(cursor.getValue());
                                    filesUtilization.put(fileAddress, new MutableLong(freeBytes));
                                }
                            }
                            synchronized (UtilizationProfile.this.filesUtilization) {
                                UtilizationProfile.this.filesUtilization.clear();
                                UtilizationProfile.this.filesUtilization.putAll(filesUtilization);
                            }
                            estimateFreeBytesAndWakeGcIfNecessary();
                        }
                    }
                });
            }
        }
    }

    /**
     * Saves utilization profile in internal store within specified transaction.
     */
    public void save(@NotNull final Transaction txn) {
        if (isDirty) {
            final StoreImpl store = env.openStore(GarbageCollector.UTILIZATION_PROFILE_STORE_NAME,
                    StoreConfig.WITHOUT_DUPLICATES, txn);
            // clear entries for already deleted files
            try (Cursor cursor = store.openCursor(txn)) {
                while (cursor.getNext()) {
                    final long fileAddress = LongBinding.compressedEntryToLong(cursor.getKey());
                    final boolean fileIsDeleted;
                    synchronized (filesUtilization) {
                        fileIsDeleted = !filesUtilization.containsKey(fileAddress);
                    }
                    if (fileIsDeleted) {
                        cursor.deleteCurrent();
                    }
                }
            }
            // save profile of up-to-date files
            final List<Map.Entry<Long, MutableLong>> filesUtilization;
            synchronized (UtilizationProfile.this.filesUtilization) {
                filesUtilization = new ArrayList<>(UtilizationProfile.this.filesUtilization.entrySet());
            }
            for (final Map.Entry<Long, MutableLong> entry : filesUtilization) {
                store.put(txn,
                        LongBinding.longToCompressedEntry(entry.getKey()),
                        CompressedUnsignedLongByteIterable.getIterable(entry.getValue().value));
            }
        }
    }

    public boolean isDirty() {
        return isDirty;
    }

    public void setDirty(boolean dirty) {
        isDirty = dirty;
    }

    int totalFreeSpacePercent() {
        final long totalBytes = this.totalBytes;
        return (int) (totalBytes == 0 ? 0 : ((totalFreeBytes * 100L) / totalBytes));
    }

    public int totalUtilizationPercent() {
        return 100 - totalFreeSpacePercent();
    }

    long getFileFreeBytes(final long fileAddress) {
        synchronized (filesUtilization) {
            final MutableLong freeBytes = filesUtilization.get(fileAddress);
            return freeBytes == null ? Long.MAX_VALUE : freeBytes.value;
        }
    }

    /**
     * Updates utilization profile with new expired loggables.
     *
     * @param loggables expired loggables.
     */
    void fetchExpiredLoggables(@NotNull final Iterable<ExpiredLoggableInfo> loggables) {
        long prevFileAddress = -1L;
        MutableLong prevFreeBytes = null;
        synchronized (filesUtilization) {
            for (final ExpiredLoggableInfo loggable : loggables) {
                final long fileAddress = log.getFileAddress(loggable.address);
                MutableLong freeBytes = prevFileAddress == fileAddress ? prevFreeBytes : filesUtilization.get(fileAddress);
                if (freeBytes == null) {
                    freeBytes = new MutableLong(0L);
                    filesUtilization.put(fileAddress, freeBytes);
                }
                freeBytes.value += loggable.length;
                prevFreeBytes = freeBytes;
                prevFileAddress = fileAddress;
            }
        }
    }

    void removeFile(final long fileAddress) {
        synchronized (filesUtilization) {
            filesUtilization.remove(fileAddress);
        }
    }

    void estimateTotalBytes() {
        // at first, estimate total bytes
        final long[] fileAddresses = log.getAllFileAddresses();
        final int filesCount = fileAddresses.length;
        int i = gc.getMinFileAge();
        totalBytes = filesCount > i ? (filesCount - i) * fileSize : 0;
        // then, estimate total free bytes
        long totalFreeBytes = 0;
        synchronized (filesUtilization) {
            for (; i < fileAddresses.length; ++i) {
                final MutableLong freeBytes = filesUtilization.get(fileAddresses[i]);
                totalFreeBytes += freeBytes != null ? freeBytes.value : fileSize;
            }
        }
        this.totalFreeBytes = totalFreeBytes;
    }

    Iterator<Long> getFilesSortedByUtilization(final long highFile) {
        final long[] fileAddresses = log.getAllFileAddresses();
        final long maxFreeBytes = fileSize * (long) gc.getMaximumFreeSpacePercent() / 100L;
        final PriorityQueue<Pair<Long, Long>> fragmentedFiles = new PriorityQueue<>(10, new Comparator<Pair<Long, Long>>() {
            @Override
            public int compare(Pair<Long, Long> leftPair, Pair<Long, Long> rightPair) {
                final long leftFreeBytes = leftPair.getSecond();
                final long rightFreeBytes = rightPair.getSecond();
                if (leftFreeBytes == rightFreeBytes) {
                    return 0;
                }
                return leftFreeBytes > rightFreeBytes ? -1 : 1;
            }
        });
        final long[] totalCleanableBytes = {0};
        final long[] totalFreeBytes = {0};
        synchronized (filesUtilization) {
            for (int i = gc.getMinFileAge(); i < fileAddresses.length; ++i) {
                final long file = fileAddresses[i];
                if (file < highFile && !gc.isFileCleaned(file)) {
                    totalCleanableBytes[0] += fileSize;
                    final MutableLong freeBytes = filesUtilization.get(file);
                    if (freeBytes == null) {
                        fragmentedFiles.add(new Pair<>(file, fileSize));
                        totalFreeBytes[0] += fileSize;
                    } else {
                        final long freeBytesValue = freeBytes.value;
                        if (freeBytesValue > maxFreeBytes) {
                            fragmentedFiles.add(new Pair<>(file, freeBytesValue));
                        }
                        totalFreeBytes[0] += freeBytesValue;
                    }
                }
            }
        }
        return new Iterator<Long>() {

            @Override
            public boolean hasNext() {
                return !fragmentedFiles.isEmpty() &&
                        totalFreeBytes[0] > totalCleanableBytes[0] * gc.getMaximumFreeSpacePercent() / 100L;
            }

            @Override
            public Long next() {
                final Pair<Long, Long> pair = fragmentedFiles.poll();
                totalFreeBytes[0] -= pair.getSecond();
                return pair.getFirst();
            }

            @Override
            public void remove() {
                throw new UnsupportedOperationException("remove");
            }
        };
    }

    /**
     * Loads utilization profile from file.
     *
     * @param path external file with utilization info in the format as created by the {@code "-d"} option
     *             of the {@code Reflect} tool
     * @see EnvironmentConfig#setGcUtilizationFromFile(String)
     */
    public void loadUtilizationFromFile(@NotNull final String path) {
        gc.getCleaner().getJobProcessor().queueAt(new Job() {
            @Override
            protected void execute() throws Throwable {
                final LongHashMap<Long> usedSpace = new LongHashMap<>();
                try {
                    try (Scanner scanner = new Scanner(new File(path))) {
                        while (scanner.hasNextLong()) {
                            final long address = scanner.nextLong();
                            final Long usedBytes = scanner.nextLong();
                            usedSpace.put(address, usedBytes);
                        }
                    }
                } catch (Throwable t) {
                    GarbageCollector.loggingError("Failed to load utilization from " + path, t);
                }
                // if an error occurs during reading the file, then GC will be too pessimistic, i.e. it will clean
                // first the files which are missed in the utilization profile.
                setUtilization(usedSpace);
            }
        }, gc.getStartTime());
    }

    /**
     * Reloads utilization profile.
     */
    public void computeUtilizationFromScratch() {
        gc.getCleaner().getJobProcessor().queueAt(new ComputeUtilizationFromScratchJob(this), gc.getStartTime());
    }

    private void clearUtilization() {
        synchronized (filesUtilization) {
            filesUtilization.clear();
        }
    }

    private void setUtilization(LongHashMap<Long> usedSpace) {
        synchronized (filesUtilization) {
            filesUtilization.clear();
            for (final Map.Entry<Long, Long> entry : usedSpace.entrySet()) {
                filesUtilization.put(entry.getKey(), new MutableLong(fileSize - entry.getValue()));
            }
        }
        estimateFreeBytesAndWakeGcIfNecessary();
    }

    private void estimateFreeBytesAndWakeGcIfNecessary() {
        estimateTotalBytes();
        if (gc.isTooMuchFreeSpace()) {
            gc.wake();
        }
    }

    /**
     * Is used instead of {@linkplain Long} for saving free bytes per file in  order to update the value in-place, so
     * reducing number of lookups in the {@linkplain #filesUtilization LongHashMap}.
     */
    private static class MutableLong {

        private long value;

        MutableLong(final long value) {
            this.value = value;
        }

        @Override
        public String toString() {
            return Long.toString(value);
        }
    }

    private static class ComputeUtilizationFromScratchJob extends Job {

        private final WeakReference<UtilizationProfile> up;

        private ComputeUtilizationFromScratchJob(@NotNull final UtilizationProfile up) {
            this.up = new WeakReference<>(up);
        }

        @Override
        protected void execute() throws Throwable {
            final UtilizationProfile up = this.up.get();
            if (up == null) return;
            final LongHashMap<Long> usedSpace = new LongHashMap<>();
            final EnvironmentImpl env = up.env;
            env.executeInReadonlyTransaction(new TransactionalExecutable() {
                @Override
                public void execute(@NotNull Transaction txn) {
                    final Log log = up.log;
                    for (final String storeName : env.getAllStoreNames(txn)) {
                        final StoreImpl store = env.openStore(storeName, StoreConfig.USE_EXISTING, txn);
                        final LongIterator it = ((TransactionBase) txn).getTree(store).addressIterator();
                        while (it.hasNext()) {
                            final long address = it.next();
                            final RandomAccessLoggable loggable = log.read(address);
                            final Long fileAddress = log.getFileAddress(address);
                            Long usedBytes = usedSpace.get(fileAddress);
                            if (usedBytes == null) {
                                usedBytes = 0L;
                            }
                            usedBytes += loggable.length();
                            usedSpace.put(fileAddress, usedBytes);
                        }
                    }
                }
            });
            up.setUtilization(usedSpace);
        }
    }
}

File: environment/src/main/kotlin/jetbrains/exodus/gc/UtilizationProfile.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.gc

import jetbrains.exodus.bindings.LongBinding
import jetbrains.exodus.core.dataStructures.Pair
import jetbrains.exodus.core.dataStructures.hash.LongHashMap
import jetbrains.exodus.core.dataStructures.hash.PackedLongHashSet
import jetbrains.exodus.env.EnvironmentConfig
import jetbrains.exodus.env.EnvironmentImpl
import jetbrains.exodus.env.StoreConfig
import jetbrains.exodus.env.Transaction
import jetbrains.exodus.io.Block
import jetbrains.exodus.io.DataReader
import jetbrains.exodus.io.DataWriter
import jetbrains.exodus.kotlin.synchronized
import jetbrains.exodus.log.AbstractBlockListener
import jetbrains.exodus.log.CompressedUnsignedLongByteIterable
import jetbrains.exodus.log.Log
import jetbrains.exodus.tree.ExpiredLoggableCollection
import java.io.File
import java.util.*
import kotlin.math.max
import kotlin.math.min

class UtilizationProfile(private val env: EnvironmentImpl, private val gc: GarbageCollector) {

    private val log: Log = env.log
    private val fileSize = log.fileLengthBound // in bytes
    private val filesUtilization = LongHashMap<MutableLong>() // file address -> number of free bytes
    private var totalBytes: Long = 0
    private var totalFreeBytes: Long = 0
    @Volatile
    var isDirty: Boolean = false

    init {
        log.addBlockListener(object : AbstractBlockListener() {
            override fun blockCreated(block: Block, reader: DataReader, writer: DataWriter) {
                filesUtilization.synchronized {
                    this[block.address] = MutableLong(0L)
                }
                estimateTotalBytes()
            }
        })
    }

    internal fun clear() = filesUtilization.synchronized { clear() }.apply { estimateTotalBytes() }

    /**
     * Loads utilization profile.
     */
    fun load() {
        val ec = env.environmentConfig
        if (ec.gcUtilizationFromScratch) {
            computeUtilizationFromScratch()
        } else {
            val storedUtilization = ec.gcUtilizationFromFile
            if (!storedUtilization.isEmpty()) {
                loadUtilizationFromFile(storedUtilization)
            } else {
                env.executeInReadonlyTransaction { txn ->
                    if (!env.storeExists(GarbageCollector.UTILIZATION_PROFILE_STORE_NAME, txn)) {
                        if (env.allStoreCount == 0L && log.numberOfFiles <= 1) {
                            clearUtilization()
                        } else {
                            computeUtilizationFromScratch()
                        }
                    } else {
                        val filesUtilization = LongHashMap<MutableLong>()
                        val store = env.openStore(GarbageCollector.UTILIZATION_PROFILE_STORE_NAME, StoreConfig.WITHOUT_DUPLICATES, txn)
                        store.openCursor(txn).use { cursor ->
                            while (cursor.next) {
                                val fileAddress = LongBinding.compressedEntryToLong(cursor.key)
                                val freeBytes = CompressedUnsignedLongByteIterable.getLong(cursor.value)
                                // don't update utilization of files being reset but not deleted
                                // if they were not actually deleted they will be collected first
                                if (freeBytes != 0L) {
                                    filesUtilization[fileAddress] = MutableLong(freeBytes)
                                }
                            }
                        }

                        // check of saved utilization is consistent, at least if it contains same files as the log
                        var inconsistent = false
                        with(PackedLongHashSet(filesUtilization.keys)) {
                            log.allFileAddresses.forEach {
                                if (!remove(it)) {
                                    inconsistent = true
                                    return@with
                                }
                            }
                            if (isNotEmpty()) {
                                inconsistent = true
                            }
                        }

                        if (inconsistent) {
                            computeUtilizationFromScratch()
                        } else {
                            this@UtilizationProfile.filesUtilization.synchronized {
                                clear()
                                putAll(filesUtilization)
                            }
                            estimateTotalBytesAndWakeGcIfNecessary()
                        }
                    }
                }
            }
        }
    }

    /**
     * Saves utilization profile in internal store within specified transaction.
     */
    fun save(txn: Transaction) {
        if (isDirty) {
            val store = env.openStore(GarbageCollector.UTILIZATION_PROFILE_STORE_NAME,
                    StoreConfig.WITHOUT_DUPLICATES, txn)
            // clear entries for already deleted files
            store.openCursor(txn).use { cursor ->
                while (cursor.next) {
                    val fileAddress = LongBinding.compressedEntryToLong(cursor.key)
                    if (filesUtilization.synchronized { !containsKey(fileAddress) }) {
                        cursor.deleteCurrent()
                    }
                }
            }
            // save profile of up-to-date files
            filesUtilization.synchronized {
                entries.forEach { entry ->
                    store.put(txn,
                        LongBinding.longToCompressedEntry(entry.key),
                        CompressedUnsignedLongByteIterable.getIterable(entry.value.value))
                }
            }
        }
    }

    fun forceSave(txn: Transaction) {
        isDirty = true
        save(txn)
    }

    internal fun totalFreeSpacePercent(): Int {
        val totalBytes = this.totalBytes
        return (if (totalBytes == 0L) 0 else totalFreeBytes * 100L / totalBytes).toInt()
    }

    fun totalUtilizationPercent(): Int {
        return 100 - totalFreeSpacePercent()
    }

    internal fun getFileFreeBytes(fileAddress: Long) = filesUtilization.synchronized {
        this[fileAddress]?.value ?: Long.MAX_VALUE
    }

    /**
     * Updates utilization profile with new expired loggables.
     *
     * @param loggables expired loggables.
     */
    internal fun fetchExpiredLoggables(loggables: ExpiredLoggableCollection) {
        var prevFileAddress = -1L
        var prevFreeBytes: MutableLong? = null
        val set = PackedLongHashSet()
        filesUtilization.synchronized {
            var l: ExpiredLoggableCollection? = loggables
            while (l != null) {
                l = l.forEach { address, length ->
                    if (set.add(address)) {
                        val fileAddress = log.getFileAddress(address)
                        val freeBytes =
                                (if (prevFileAddress == fileAddress) prevFreeBytes else this[fileAddress])
                                        ?: MutableLong(0L).also { this[fileAddress] = it }
                        freeBytes.value += length.toLong()
                        prevFreeBytes = freeBytes
                        prevFileAddress = fileAddress
                    }
                }
            }
        }
    }

    internal fun removeFile(fileAddress: Long): Unit = filesUtilization.synchronized { remove(fileAddress) }

    internal fun resetFile(fileAddress: Long) = filesUtilization.synchronized { this[fileAddress]?.value = 0L }

    internal fun estimateTotalBytes() {
        val fileAddresses = log.allFileAddresses
        val filesCount = fileAddresses.size
        val minFileAge = gc.minFileAge
        val totalFreeBytes: Long = filesUtilization.synchronized {
            (minFileAge until filesCount).fold(0L) { sum, i ->
                sum + (this[fileAddresses[i]]?.value ?: fileSize)
            }
        }
        val totalBytes = if (filesCount > minFileAge) (filesCount - minFileAge) * fileSize else 0
        this.totalBytes = totalBytes
        this.totalFreeBytes = min(totalFreeBytes, totalBytes)
    }

    internal fun getFilesSortedByUtilization(highFile: Long): Iterator<Long> {
        val fileAddresses = log.allFileAddresses
        val maxFreeBytes = fileSize * gc.maximumFreeSpacePercent.toLong() / 100L
        val fragmentedFiles = PriorityQueue(10, Comparator<Pair<Long, Long>> { leftPair, rightPair ->
            val leftFreeBytes = leftPair.second
            val rightFreeBytes = rightPair.second
            if (leftFreeBytes == rightFreeBytes) {
                return@Comparator 0
            }
            if (leftFreeBytes > rightFreeBytes) -1 else 1
        })
        var totalCleanableBytes = 0L
        var totalFreeBytes = 0L
        filesUtilization.synchronized {
            (gc.minFileAge until fileAddresses.size).forEach { i ->
                val file = fileAddresses[i]
                if (file < highFile && !gc.isFileCleaned(file)) {
                    totalCleanableBytes += fileSize
                    val freeBytes = this[file]
                    totalFreeBytes += if (freeBytes == null) {
                        fragmentedFiles.add(Pair(file, fileSize))
                        fileSize
                    } else {
                        val freeBytesValue = freeBytes.value
                        if (freeBytesValue > maxFreeBytes) {
                            fragmentedFiles.add(Pair(file, freeBytesValue))
                        }
                        freeBytesValue
                    }
                }
            }
        }
        return object : Iterator<Long> {

            override fun hasNext(): Boolean {
                return !fragmentedFiles.isEmpty() && totalFreeBytes > totalCleanableBytes * gc.maximumFreeSpacePercent / 100L
            }

            override fun next(): Long {
                val pair = fragmentedFiles.poll()
                totalFreeBytes -= pair.second
                return pair.first
            }
        }
    }

    /**
     * Loads utilization profile from file.
     *
     * @param path external file with utilization info in the format as created by the `"-d"` option
     * of the `Reflect` tool
     * @see EnvironmentConfig.setGcUtilizationFromFile
     */
    fun loadUtilizationFromFile(path: String) {
        gc.cleaner.getJobProcessor().queueAt(GcJob(gc) {
            val usedSpace = LongHashMap<Long>()
            try {
                Scanner(File(path)).use { scanner ->
                    while (scanner.hasNextLong()) {
                        val address = scanner.nextLong()
                        val usedBytes = scanner.nextLong()
                        usedSpace[address] = usedBytes
                    }
                }
            } catch (t: Throwable) {
                GarbageCollector.loggingError(t) { "Failed to load utilization from $path" }
            }

            // if an error occurs during reading the file, then GC will be too pessimistic, i.e. it will clean
            // first the files which are missed in the utilization profile.
            filesUtilization.synchronized {
                clear()
                setUtilization(usedSpace)
            }
        }, gc.startTime)
    }

    /**
     * Reloads utilization profile.
     */
    fun computeUtilizationFromScratch() {
        GarbageCollector.loggingInfo { "Queueing ComputeUtilizationFromScratchJob" }
        gc.cleaner.getJobProcessor().queueAt(ComputeUtilizationFromScratchJob(gc), gc.startTime)
    }

    internal fun estimateTotalBytesAndWakeGcIfNecessary() {
        estimateTotalBytes()
        if (gc.isTooMuchFreeSpace) {
            gc.wake()
        }
    }

    internal fun setUtilization(usedSpace: LongHashMap<Long>) = filesUtilization.synchronized {
        for ((fileAddress, usedBytes) in usedSpace) {
            (this[fileAddress] ?: MutableLong(0L).also { this[fileAddress] = it }).value += max((fileSize - usedBytes), 0L)
        }
    }

    private fun clearUtilization() = filesUtilization.synchronized { clear() }


    /**
     * Is used instead of [Long] for saving free bytes per file in  order to update the value in-place, so
     * reducing number of lookups in the [LongHashMap][.filesUtilization].
     */
    private class MutableLong constructor(var value: Long) {

        override fun toString(): String {
            return value.toString()
        }
    }
}